core/src/main/scala/org/broadinstitute/dsde/rawls/expressions/OutputExpression.scala

```diff
@@ -10,22 +10,27 @@ sealed trait OutputExpressionTarget {
 case object ThisEntityTarget extends OutputExpressionTarget { override val prefix = "this." }
 case object WorkspaceTarget extends OutputExpressionTarget { override val prefix = "workspace." }
 
-case class OutputExpression(target: OutputExpressionTarget, attributeName: AttributeName)
+sealed trait OutputExpression
+case object UnboundOutputExpression extends OutputExpression
+case class BoundOutputExpression(target: OutputExpressionTarget, attributeName: AttributeName, attribute: Attribute) extends OutputExpression
```

Comment:
If I understand correctly: this represents `this.foo = 2`, where `target` is `this`, `attributeName` is `foo`, and `attribute` is `2`?

I suggest renaming `OutputExpressionTarget.prefix` to `root` and possibly `attribute` to `value` or somesuch.

I think the `remainder` function should live in a companion object of `BoundOutputExpression`, at which point you could call it `stripTarget` or `generateAttributeName` or something. There's something weird to me about passing an expression to a target you just generated in order to strip the target off the expression. It's both redundant and feels error-prone (what if you accidentally pass a `this.` expression to a `WorkspaceTarget`?).